File: ios/RNInputKit/Health/HealthBridge.m
//
//  HealthBridge.m
//  InputKitProber
//
//  Created by Tatsuya Kaneko on 23/05/2017.
//  Copyright © 2017 Facebook. All rights reserved.
//

#import <React/RCTBridgeModule.h>
#import <React/RCTConvert.h>
//#import "InputKitProber-Swift.h"

@interface RCT_EXTERN_MODULE(HealthBridge, NSObject)

RCT_EXTERN_METHOD(isAvailable:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(isProviderInstalled:(NSString * _Nonnull)providerName resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(requestPermissions:(NSArray<NSString *> * _Nonnull)typesToRead resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(getStepCount:(NSDate * _Nonnull)startDate endDate:(NSDate * _Nonnull)endDate resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(getDistance:(NSDate * _Nonnull)startDate endDate:(NSDate * _Nonnull)endDate resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(getDistanceSamples:(NSDate * _Nonnull)startDate endDate:(NSDate * _Nonnull)endDate limit:(NSInteger * _Nonnull)limit resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(getStepCountDistribution:(NSDate * _Nonnull)startDate endDate:(NSDate * _Nonnull)endDate interval:(NSString * _Nonnull)interval resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(getAccurateDistance:(NSDate * _Nonnull)startDate endDate:(NSDate * _Nonnull)endDate resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(getSleepAnalysisSamples:(NSDate * _Nonnull)startDate endDate:(NSDate * _Nonnull)endDate resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(startMonitoring:(NSString * _Nonnull)typeString resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(stopMonitoring:(NSString * _Nonnull)typeString resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

// Tracking Features using CoreMotion

RCT_EXTERN_METHOD(startTracking:(NSString * _Nonnull)type startDate: (NSDate * _Nonnull)startDate resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(stopTracking:(NSString * _Nonnull)type resolve: (RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(getHistoricActivityTrackingData:(NSDate * _Nonnull)startDate endDate:(NSDate * _Nonnull)endDate resolve:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);

RCT_EXTERN_METHOD(stopTrackingAll:(RCTPromiseResolveBlock _Nonnull)resolve reject:(RCTPromiseRejectBlock _Nonnull)reject);


+ (BOOL) requiresMainQueueSetup {
  return false;
}

@end

File: ios/RNInputKit/Health/HealthBridge.swift
//
//  HealthBridge.swift
//  Erasmus
//
//  Created by Tatsuya Kaneko on 19/06/2017.
//  Copyright © 2017 Sense Health BV. All rights reserved.
//

import Foundation
import HealthKit
import CoreMotion

@objc (HealthBridge)
class HealthBridge: RCTEventEmitter { // No NSObject inheritance, since RCTEventEmitter inherits from NSObject already
  
  let healthStore = HKHealthStore()
  
  // For realtime tracking
  let pedoMeter = CMPedometer()
  var currentTrackedTypes = [String]()

  @objc override func supportedEvents() -> [String]! {
    return [SupportedEvents.inputKitUpdates]
  }
  
  static let quantiyTypes: [String: HKSampleType] = [
    "stepCount": HKObjectType.quantityType(forIdentifier: .stepCount)!,
    "distanceWalkingRunning": HKObjectType.quantityType(forIdentifier: .distanceWalkingRunning)!
  ]
  
  // Default units for quantity types
  static let quantityUnits: [HKQuantityTypeIdentifier: HKUnit] = [
    HKQuantityTypeIdentifier.stepCount: HKUnit.count(),
    HKQuantityTypeIdentifier.distanceWalkingRunning: HKUnit.meter()
  ]
  
  static let categoryTypes: [String: HKSampleType] = [
    "sleep": HKObjectType.categoryType(forIdentifier: .sleepAnalysis)!,
  ]
  
  // It would be a time saver if we could expose Health.type to RN side, but unfortunately we can not.
  // We could convert string to HKObject directly here, but Bridge should not know too much detail about HealthKit. Rather Health class in InputKit should know more details. Therefore, we use mapping from string to custom type Health.type.
  static let readPermissions: [String: HKSampleType] = quantiyTypes.merged(with: categoryTypes)
  
  
  @objc func isAvailable(_ resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    resolve(HKHealthStore.isHealthDataAvailable())
  }

  @objc func isProviderInstalled(_ providerName: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard !providerName.isEmpty else {
      reject(kInputKitError, "Provider name must be provided.", nil)
      return
    }
    
    //TODO: add another health provider app installation check
    if providerName == ProviderName.healthKit {
      //Consider that HealthKit has been built in iOS devices, we can simply resolve this request
      resolve(true)
      return
    }
    
    reject(kInputKitError, providerName + " is not supported in InputKit!", nil)
  }
  
  @objc func requestPermissions(_ typesToRead: [String], resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock ) {
    let typeSet = Set(typesToRead.compactMap { HealthBridge.readPermissions[$0] })
    
    healthStore.requestAuthorization(toShare: nil, read: typeSet) {
      (success, error) in
      
      guard error == nil, success == true else {
        reject(kHealthKitError, "Error while requesting permissions", error)
        return
      }
      
      resolve(nil)
    }
  }

  func emitEvent(withName name: String, body: [String : Any], completion: @escaping () -> Void) {
    guard let eventHandler = self.bridge.module(forName: "EventHandlerBridge") as? EventHandlerBridge else {
      //TODO: this should never happen
      NativeLogger.logNativeEvent("EventHandlerBridge doesn't exist in bridge.module")
      return
    }
    
    NativeLogger.logNativeEvent("\(NSStringFromClass(type(of: self)))::\(#function)")
    eventHandler.emit(withName: name, body: body, completion: completion)
  }
}
